raft/storage.go

```diff
@@ -57,6 +57,17 @@ type Storage interface {
 	Snapshot() (pb.Snapshot, error)
 }
 
+// InitialState is the return value of the Storage.InitialState method.
+type InitialState struct {
+	HardState pb.HardState
+	ConfState pb.ConfState
+
+	// AppliedIndex is the last index which has been applied to the state machine,
+	// for implementations which persist the state machine to disk. If the state
+	// machine is not persistent, this field may be zero.
+	AppliedIndex uint64
```

Comment:
OK, we can work with that. Do we need to consider any other interaction between the applier/state machine and raft, or is the initial applied index the only thing we need? 

For reference, we've had a version of this conversation before in cockroachdb/etcd#6